### PR TITLE
Add --preview-dart-2 to 'flutter test'

### PR DESCRIPTION
**Pull request by aam**
This is toward fixing https://github.com/flutter/flutter/issues/14105.

This relies on https://github.com/flutter/engine/pull/4559.

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 1]
```diff
@@ -191,14 +197,28 @@ class _FlutterPlatform extends PlatformPlugin {
...
 
       // Start the engine subprocess.
-      printTrace('test $ourTestCount: starting shell process');
+      printTrace('test $ourTestCount: starting shell process${previewDart2? " in preview-dart-2 mode":""}');
+
+      final String mainDart = previewDart2
+        ? await compile(
+            sdkRoot: artifacts.getArtifactPath(Artifact.flutterPatchedSdkPath),
+            incrementalCompilerByteStorePath: '' /* not null is enough */,
+            mainPath: listenerFile.path,
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: also add parameter strongMode: true

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 2]
```diff
@@ -191,14 +197,28 @@ class _FlutterPlatform extends PlatformPlugin {
...
+            packagesPath: PackageMap.globalPackagesPath
+          )
+        : listenerFile.path;
+
+      // bundlePath needs to point to a folder with `platform.dill` file.
+      final String bundlePath = previewDart2 ? artifacts.getArtifactPath(Artifact.flutterPatchedSdkPath) : null;
+
       final Process process = await _startProcess(
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: The --strong option needs to be passed to the engine being invoked also as we will want to also test the runtime strong mode checks.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 1 above)
Comment (aam) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: The test fails in strongMode, we won't be able to confirm the flow until flutter and frontend is strong mode clean. Given that I would prefer to add strong mode separately in follow up PR.

Comment (a-siva) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Ok as discussed off line you will land this PR, follow it up with a PR to make it work on linux too and another PR to add the strong mode option.
It is important to get the strong mode option is as we want to get to the errors quickly.

**Review by a-siva — APPROVED** [review]

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 3]
```diff
@@ -492,6 +513,9 @@ void main() {
     }
     if (host.type == InternetAddressType.IP_V6)
       command.add('--ipv6');
+    if (bundlePath != null) {
+      command.add('--flutter-assets-dir=$bundlePath');
```
Comment (mehmetf) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: @aam Do you recall why this was needed?

It is a bit odd that we are passing the SDK path as the asset dir.

Comment (mehmetf) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: This implies all assets to be used by tester needs to be put in the sdk path. That's quite odd. In platform packaging the dill files are not found in the asset directory. Would it make sense to separate these concerns and change the way the dill files are located in the tester? [probably a new flag]?

Comment (aam) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Yeah, it does make sense.

Comment (mehmetf) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: @tvolkert FYI I will attempt this if you don't have any objection. The request is to change tester to locate dill files w/o needing the asset flag so we can safely remove this particular line of code in flutter_platform.dart. When we add proper asset support [a separate param to installHook?], we can add it back. If you disagree with the methodology, let me know.

**Pull request merged**